### PR TITLE
C++: Work around two false positive issues with the UnusedLocals.ql query

### PR DESCRIPTION
**Pull request by geoffw0**
Test cases and workaround fixes for the issues highlighted by https://github.com/github/codeql/issues/4569 and https://github.com/github/codeql/issues/4570.

Sample of results from LGTM that will now be excluded: https://lgtm.com/query/6496877322277687392/

---

@igfoo I believe both cases may be caused by limitations in the extractor where a variable is accessed in a particular context, but no `VariableAccess` in generated.
 - on line 143 of the test; the expression on the left side of the `;` (a `ConditionDeclExpr`) appears not to have been extracted correctly.  If this is fixed I would like to remove the exclusion for `ConditionDeclExpr` from this query (affects up to ~0.2% of functions).
 - on line 164 of the test; ideally there would be a `VariableAccess` inside the expression that is currently an 'unknown literal', so we don't need to exclude all functions containing unknown literals from the test (affects up to ~10% of functions).

Are you happy to deal with these, should I create issues, or are they already on your radar?

### CODE REVIEW

**Comment by igfoo** [comment]
I had a quick look at the first one, and I can't immediately see what's going wrong. Could you file a couple of tickets please? Thanks!

File: cpp/change-notes/2020-11-02-unused-local-variable.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+lgtm,codescanning
+* Two issues causing the 'Unused local variable' (cpp/unused-local-variable) to produce false positive results have been fixed.
```
Comment (jbj) [inline cpp/change-notes/2020-11-02-unused-local-variable.md:2]:
```suggestion
* Two issues causing the 'Unused local variable' query (`cpp/unused-local-variable`) to produce false positive results have been fixed.
```

File: cpp/ql/src/Best Practices/Unused Entities/UnusedLocals.ql [hunk 2]
```diff
@@ -57,5 +57,10 @@ where
   not declarationHasSideEffects(v) and
   not exists(AsmStmt s | f = s.getEnclosingFunction()) and
   not v.getAnAttribute().getName() = "unused" and
-  not any(ErrorExpr e).getEnclosingFunction() = f // unextracted expr likely used `v`
+  not any(ErrorExpr e).getEnclosingFunction() = f and // unextracted expr may use `v`
+  not exists(Literal l |
+    l.getEnclosingFunction() = f and
+    not exists(l.getValue())
+  ) and
+  not any(ConditionDeclExpr cde).getEnclosingFunction() = f
```
Comment (jbj) [inline cpp/ql/src/Best Practices/Unused Entities/UnusedLocals.ql:65]: Please add some comments that'll help us know when to delete these extra conditions in the future.

Comment (geoffw0) [reply cpp/ql/src/Best Practices/Unused Entities/UnusedLocals.ql]: Done.  The details aren't really fresh in my mind any more, but at least its documented that they should at some point be removed.

**Comment by geoffw0** [comment]
> Could you file a couple of tickets please?

Done: https://github.com/github/codeql-c-extractor-team/issues/150 and https://github.com/github/codeql-c-extractor-team/issues/151.

It's a while since I looked at this so apologies for any inaccuracies - particularly on the second one.

**Comment by jbj** [comment]
The qlformat test failed

**Comment by geoffw0** [comment]
Thanks, fixed.

**Review by jbj — APPROVED** [review]

**Pull request merged**